run.py

```diff
@@ -33,17 +37,80 @@ def _get_opt_parser(self):
+    def _get_connection(self, hostname, username, password):
         client = paramiko.SSHClient()
         client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
-        client.connect(hostname=self.options.host, username=self.options.user,
-                       password=self.options.password)
+        client.connect(hostname=hostname, username=username,
+                       password=password)
         return client
 
-    def _execute_command(self, command):
-        (stdin, stdout, stderr) = self.connection.exec_command(command)
+    def _execute_command(self, connection, command):
+        (stdin, stdout, stderr) = connection.exec_command(command)
         return (stdout, stderr)
 
+    def troubleshoot_mon(self):
+        command = 'sudo ceph health'
+        (output, err) = self._execute_command(self.connection, command)
+
+        try:
+            self._get_eof(output, command)
+            cluster_status = output.read().split(' ')[0].strip()
+
+        except TimeoutError as err:
+            # cluster is screwed
+            # Best guess some mon machines are down
+            # just printing error for now
+            # TODO
+            print err
+            return
+
+        if cluster_status in self.GOOD_HEALTH:
+            print cluster_status
+            print 'All good up here, no mon issues :-)'
+
+        if cluster_status in self.BAD_HEALTH:
+            print 'MON Status : ', cluster_status
+            print '\nProbable cause Ceph service not running in some machines'
+            print 'Try & start service in the machines not in quorum',
+            print '(yes/no) (default no)?',
+            response = raw_input()
+
+            if response in ['no', '']:
+                print 'not proceeding with updating machines, aborting'
+                return
+            if response not in ['yes']:
+                print 'response not valid, aborting'
+                return
+
+            command = 'sudo ceph mon_status'
+            (output, err) = self._execute_command(self.connection, command)
+
+            mon_status = eval(output.read())
+            quorum_list = mon_status['quorum']
+            mon_list = mon_status['monmap']['mons']
+
+            if len(quorum_list) != len(mon_list):
+                for mon in mon_list:
+                    if mon['rank'] not in quorum_list:
+                        print '\n' + mon['name'] + ' not in quorum list,',
+                        print 'restarting ceph services'
+                        mon_addr = mon['addr'].split(':')[0]
+
+                        self._restart_ceph_services(mon_addr,
+                                                    self.options.user,
+                                                    self.options.password)
+
+    def _restart_ceph_services(self, addr, username, password):
+        connection = self._get_connection(addr, username, password)
+        self._execute_command(connection, 'sudo start ceph-all')
```

Comment:
It's easy to see what systemd ceph looks like. Just deploy the xenial ceph
with juju
On May 18, 2016 6:46 PM, "Kumar Rishabh" <notifications@github.com> wrote:

> In run.py
> <https://github.com/CanonicalLtd/DiagCephCluster/pull/1#discussion_r63811918>
> :
>
> > +            mon_list = mon_status['monmap']['mons']
> > +
> > +            if len(quorum_list) != len(mon_list):
> > +                for mon in mon_list:
> > +                    if mon['rank'] not in quorum_list:
> > +                        print '\n' + mon['name'] + ' not in quorum list,',
> > +                        print 'restarting ceph services'
> > +                        mon_addr = mon['addr'].split(':')[0]
> > +
> > +                        self._restart_ceph_services(mon_addr,
> > +                                                    self.options.user,
> > +                                                    self.options.password)
> > +
> > +    def _restart_ceph_services(self, addr, username, password):
> > +        connection = self._get_connection(addr, username, password)
> > +        self._execute_command(connection, 'sudo start ceph-all')
>
> Yeah, almost all the systems I work on are sysvinit based and hence didnt
> check. Will correct that.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/CanonicalLtd/DiagCephCluster/pull/1/files/59a5b986b3412dbddf69c81ff5df6ff2c8b9e297#r63811918>
>